Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/ProgressIndicators/ProgressIndicatorLoadingBar.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading.Tasks;
+using TMPro;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.SceneTransitions
+{
+    /// <summary>
+    /// Displays progress bar and (optionally) percentage text.
+    /// </summary>
+    [ExecuteAlways]
+    public class ProgressIndicatorLoadingBar : MonoBehaviour, IProgressIndicator
+    {
+        const float SmoothProgressSpeed = 0.25f;
+
+        public Transform MainTransform { get { return transform; } }
+        public ProgressIndicatorState State { get { return state; } }
+        public float Progress { set { targetProgress = value; } }
+        public string Message { set { messageText.text = value; } }
+
+        // The animated progress bar object
+        [SerializeField]
+        private Transform progressBar = null;
+
+        // The message text used by the 'Visible' message style
+        [SerializeField]
+        private TextMeshPro messageText = null;
+
+        // The progress text used by all non-'None' progress styles
+        [SerializeField]
+        private TextMeshPro progressText = null;
+
+        [Tooltip("Format to use when converting progress number to string.")]
+        [SerializeField]
+        private string progressStringFormat = "{0:P0}";
+
+        [Tooltip("Whether to display the percentage as text in addition to the loading bar.")]
+        [SerializeField]
+        private bool displayPercentage = true;
+        
+        private float smoothProgress = 0f;
+        private ProgressIndicatorState state = ProgressIndicatorState.Closed;
+        private Vector3 barScale = Vector3.one;
+
+        /// <summary>
+        /// Target for smooth progress, 0-1. Serialized to test in editor.
+        /// </summary>
+        [SerializeField]
+        [Range(0f, 1f)]
+        private float targetProgress = 0f;
+        
+        public async Task OpenAsync()
+        {
+            gameObject.SetActive(true);
+
+            state = ProgressIndicatorState.Opening;
+
+            await Task.Yield();
+
+            state = ProgressIndicatorState.Open;
+        }
+
+        public async Task CloseAsync()
+        {
+            state = ProgressIndicatorState.Closing;
+
+            await Task.Yield();
```

Comment:
It might be worth documenting that OpenAsync and CloseAsync are not thread-safe and one should be fully-awaited before the other one is called. Otherwise it would be too easy to get into an inconsistent state (i.e. if you call open and then close but don't await either, it's possible that you'll end up in an Open state even when the gameObject is set to not active).

This does put the burden on the caller, which maybe is fine, but if we do that it should be documented.

Reply:
Yikes good point. These were designed for use by a system that wouldn't have made this mistake. Good catch.